urldeb33fd2fa_friedemann_distripedia_tgz-pJ8-org_json_TestJ8/src/websocket/server/DistripediaWorker.java

```diff
@@ -0,0 +1,504 @@
+
+import websocket.common.Logger;
+import websocket.common.Message;
+import websocket.common.Util;
+
+public class DistripediaWorker extends WebSocket implements Runnable
+{
+	// holds all article titles that the server assumes are stored on the client
+	private final Set<String> clientArticles = new HashSet<String>();
+
+	// holds the stratus client if the client is a stratus client (256bit hash)
+	private String stratusPeerId = null;
+
+	// this objects id
+	private int id = -1;
+	
+	private String name = "null";
+
+	public void setId(int id)
+	{
+		this.id = id;
+	}
+
+	public String getName()
+	{
+		return name;
+	}
+	
+	public void setName(String name)
+	{
+		this.name = name;
+	}
+
+	private boolean hasStratusClient()
+	{
+		return stratusPeerId != null;
+	}
+
+	// serves a client at given socket
+	public DistripediaWorker(final Socket socket) throws IOException
+	{
+		super(socket);
+	}
+
+	public void run()
+	{
+		try
+		{
+			// send name to client
+			sendClientName();
+			
+			// method blocks until something bad happens
+			listenForMessages();
+
+			// not listening anymore, close streams/connection
+			close();
+
+			// if we reach this point the client has (been) disconnected
+			log("client disconnected.");
+			logActivity(108, getName(), "Server", hasStratusClient(), "client logging off");
+
+			// remove all links to articles
+			removeAllArticleLinks();
+		} catch (Exception e)
+		{
+			Logger.log("SEVERE! distripedia worker crash!");
+			e.printStackTrace();
+		}
+	}
+
+	// removes all links to the articles that the client has from the server
+	// hashmaps
+	private void removeAllArticleLinks()
+	{
+		for (String article : clientArticles)
+		{
+			// remove articles from hashmap for all workers
+			removeArticleLinkFromMap(article, WebSocketServer.allWorkersWithArticles);
+
+			// also remove from hashmap for stratus clients if client was one
+			if (this.hasStratusClient())
+			{
+				removeArticleLinkFromMap(article, WebSocketServer.stratusWorkersWithArticles);
+			}
+		}
+
+		// empty the set now
+		clientArticles.clear();
+
+		// Logger.log("article/worker hash map information:");
+		// Logger.log("| stratus: " + WebSocketServer.stratusWorkersWithArticles.toString());
+		// Logger.log("| all: " + WebSocketServer.allWorkersWithArticles.toString());
+	}
+
+	// removes an article from from the server hashmap
+	private void removeArticleLinkFromMap(String article, Map<String, List<DistripediaWorker>> map)
+	{
+		List<DistripediaWorker> articleList = map.get(article);
+
+		synchronized (articleList)
+		{
+			articleList.remove(this);
+		}
+	}
+
+	// adds one article and this worker instance to a server hashmap
+	private void addArticleLinkToMap(String article,
+			ConcurrentHashMap<String, List<DistripediaWorker>> map)
+	{
+		List<DistripediaWorker> articleList = map.get(article);
+
+		if (articleList == null)
+		{
+			articleList = new ArrayList<DistripediaWorker>();
+			map.put(article, articleList);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **128, 130, and 133**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and uses that reference at line: **140**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.